crates/engine/tree/src/tree/mod.rs

```diff
@@ -2275,177 +2262,162 @@ where
-                self.config.use_state_root_task()
-            {
-                let consistent_view = ConsistentDbView::new_with_latest_tip(self.provider.clone())?;
+        let (state_root_handle, state_hook) = if persistence_not_in_progress &&
+            self.config.use_state_root_task()
+        {
+            let consistent_view = ConsistentDbView::new_with_latest_tip(self.provider.clone())?;
 
-                let state_root_config = StateRootConfig::new_from_input(
-                    consistent_view.clone(),
-                    self.compute_trie_input(consistent_view.clone(), block.header().parent_hash())
-                        .map_err(|e| InsertBlockErrorKind::Other(Box::new(e)))?,
-                );
+            let state_root_config = StateRootConfig::new_from_input(
+                consistent_view.clone(),
+                self.compute_trie_input(consistent_view.clone(), block.header().parent_hash())
+                    .map_err(|e| InsertBlockErrorKind::Other(Box::new(e)))?,
+            );
 
-                let provider_ro = consistent_view.provider_ro()?;
-                let nodes_sorted = state_root_config.nodes_sorted.clone();
-                let state_sorted = state_root_config.state_sorted.clone();
-                let prefix_sets = state_root_config.prefix_sets.clone();
-
-                // context will hold the values that need to be kept alive
-                let context =
-                    StateHookContext { provider_ro, nodes_sorted, state_sorted, prefix_sets };
-
-                // it is ok to leak here because we are in a scoped thread, the
-                // memory will be freed when the thread completes
-                let context = Box::leak(Box::new(context));
-
-                let blinded_provider_factory = ProofBlindedProviderFactory::new(
-                    InMemoryTrieCursorFactory::new(
-                        DatabaseTrieCursorFactory::new(context.provider_ro.tx_ref()),
-                        &context.nodes_sorted,
-                    ),
-                    HashedPostStateCursorFactory::new(
-                        DatabaseHashedCursorFactory::new(context.provider_ro.tx_ref()),
-                        &context.state_sorted,
-                    ),
-                    context.prefix_sets.clone(),
-                );
+            let provider_ro = consistent_view.provider_ro()?;
+            let nodes_sorted = state_root_config.nodes_sorted.clone();
+            let state_sorted = state_root_config.state_sorted.clone();
+            let prefix_sets = state_root_config.prefix_sets.clone();
+
+            let trie_cursor = Arc::new(InMemoryTrieCursorFactory::new(
+                DatabaseTrieCursorFactory::new(provider_ro.tx_ref()),
+                nodes_sorted,
+            ));
+            let state_cursor = Arc::new(HashedPostStateCursorFactory::new(
+                DatabaseHashedCursorFactory::new(provider_ro.tx_ref()),
+                state_sorted,
+            ));
```

Comment:
I think this is the problematic section because this borrows and is not 'static

> 2276 |             let provider_ro = consistent_view.provider_ro()?;
     |                 ----------- binding `provider_ro` declared here
...
2286 |                 DatabaseHashedCursorFactory::new(provider_ro.tx_ref()),
     |                                                  ^^^^^^^^^^^---------
     |                                                  |
     |                                                  borrowed value does not live long enough
     |                                                  argument requires that `provider_ro` is borrowed for

Reply:
yes exactly, we previously had the same issue with `nodes_sorted` and `state_sorted`, was fixed by turning reference parameters into `Arc`s here https://github.com/paradigmxyz/reth/pull/13749/commits/a5103d506d0d84a3820dc1f74002dca534e19059, currently in progress passing ownership of the provider factory as you mentioned